### PR TITLE
SNOW-1660802 Implement dataframe groupby fillna (group by axis=0)

### PR DESCRIPTION
**Pull request by sfc-gh-evandenberg**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-1660802 Implement dataframe groupby fillna (group by axis=0)

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Add support for DataFrameGroupBy.fillna when group by axis=0.  Note that axis=1 will be depreciated post pandas 2.1.0 so not investing in that support.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/extensions/groupby_overrides.py [hunk 1]
```diff
@@ -508,9 +508,28 @@ def diff(self):
...
+        self,
+        value: Any = None,
+        method: Optional[FillnaOptions] = None,
+        axis: Optional[Axis] = None,
+        inplace: Optional[bool] = False,
+        limit: Optional[int] = None,
+        downcast: Optional[dict] = None,
+    ):
         # TODO: SNOW-1063349: Modin upgrade - modin.pandas.groupby.DataFrameGroupBy functions
-        ErrorMessage.method_not_implemented_error(name="fillna", class_="GroupBy")
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/extensions/groupby_overrides.py:513]: Do we know if `SeriesGroupBy.fillna` will be supported or not? If not, then we should still raise an error.

File: tests/integ/modin/groupby/test_groupby_fillna.py
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_fillna.py:1]: Can we add a test for SeriesGroupby?

File: docs/source/modin/supported/groupby_supported.rst [hunk 2]
```diff
@@ -106,7 +106,8 @@ Computations/descriptive stats
 +-----------------------------+---------------------------------+----------------------------------------------------+
 | ``ffill``                   | N                               |                                                    |
 +-----------------------------+---------------------------------+----------------------------------------------------+
-| ``fillna``                  | N                               |                                                    |
+| ``fillna``                  | P                               | GroupBy axis = 0 is supported.                     |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/supported/groupby_supported.rst:109]: It seems axis=1 is also supported?

Comment (sfc-gh-evandenberg) [reply docs/source/modin/supported/groupby_supported.rst:109]: No, the general case of groupby axis=1 is not supported all the df based on the docstrings.

From dataframe_supported.rst:
Screenshot 2024-09-23 at 4 10 42 PM

File: src/snowflake/snowpark/modin/plugin/extensions/groupby_overrides.py (hunk 1 above)
Comment (sfc-gh-evandenberg) [reply src/snowflake/snowpark/modin/plugin/extensions/groupby_overrides.py:513]: I added support for Series.groupby.fillna as get it from same implementation.

File: tests/integ/modin/groupby/test_groupby_fillna.py
Comment (sfc-gh-evandenberg) [reply tests/integ/modin/groupby/test_groupby_fillna.py:1]: Added test

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@
 - Added the following new functions in `snowflake.snowpark.functions`:
   - `make_interval`
 - Added support for using Snowflake Interval constants with `Window.range_between()` when the order by column is TIMESTAMP or DATE type
+- Added support for `DataFrameGroupBy.fillna`.
```
Comment (sfc-gh-helmeleegy) [inline CHANGELOG.md:12]: Can we also add `SeriesGroupBy.fillna` here?

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM, thanks Eric!

**Review by sfc-gh-nkrishna — APPROVED** [review]
Left one minor comment, but otherwise LGTM!

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 4]
```diff
@@ -5397,6 +5400,280 @@ def groupby_value_counts(
...
+                data_column_pairs = list(
+                    zip(
+                        frame.data_column_snowflake_quoted_identifiers,
+                        new_snowflake_quoted_identifiers,
+                        data_column_group_keys_mask,
+                    )
+                )
+
+                if method == "bfill":
+                    data_column_pairs.reverse()
```
[the remark this answers is not in the document]
Comment (sfc-gh-evandenberg) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:5599]: Added

**Pull request merged**